### PR TITLE
feat: Implement ATPIT

### LINKED ISSUE #16584: Support ATPIT
https://github.com/rust-lang/rust/pull/120700 is being stabilized soon

### PR DESCRIPTION
**Pull request by ShoyuVanilla**
Resolves #16584

Note: This implementation only works for ATPIT, not for TAIT.
The main hinderence that blocks the later is the defining sites of TAIT can be inner blocks like in;
```rust
type X = impl Default;

mod foo {
    fn bar() -> super::X {
        ()
    }
}
```
So, to figure out we are defining it or not, we should recursively probe for nested modules and bodies.

For ATPIT, we can just look into current body because `error[E0401]: can't use 'Self' from outer item` prevent such nested structures;

```rust
trait Foo {
    type Item;
    fn foo() -> Self::Item;
}

struct Bar;

impl Foo for Bar {
    type Item = impl Default;
    fn foo() -> Self::Item {
        fn bar() -> Self::Item {
                    ^^^^^^^^^^
                    |
                    use of `Self` from outer item
                    refer to the type directly here instead
            5
        }
        bar()
    }
}
```

But this implementation does not checks for unification of same ATPIT between different bodies, monomorphization, nor layout for similar reason. (But these can be done with lazyness if we can utilize something like "mutation of interned value" with `db`. I coundn't find such thing but I would appreciate it if such thing exists and you could let me know 😅)

### CODE REVIEW

**Comment by ShoyuVanilla** [comment]
Closed by mistake 😅

**Comment by Veykril** [comment]
TAITs I think there is still a discussion how to mark defining use sites (as without a marker like thing we won't be able to support it). I don't remember if the current consensus was on the type alias being used in a signature or if it was a marker attribute.

**Comment by ShoyuVanilla** [comment]
> That sounds horrible and unlikely to be supported by our query system.

Haha, yes, indeed 😅

**Comment by ShoyuVanilla** [comment]
To elaborate this more, to find out "What concrete type is hidden behind ATPIT opaque type" would require inference of all bodies inside the `impl Trait` block that ATPIT belongs to, but this is heavy single query is not desirable in LSP, I think.
So, I've thought about leaving them as unevaluated and if we find them as a "side effect" while querying associated items' - functions and constants - bodies, updating the inference result to unevaluated items.
But I agree with you that such "indirect" query and the methodologies for doing so - like global mutability in our query system or db storage - would be horrible and root of many problems like inconsistent states.

**Review by Veykril — APPROVED** [review]

File: crates/hir-ty/src/infer/coerce.rs [hunk 1]
```diff
@@ -276,6 +276,23 @@ impl InferenceTable<'_> {
             return success(simple(Adjust::NeverToAny)(to_ty.clone()), to_ty.clone(), vec![]);
         }
 
+        // If we are coercing into an ATPIT, coerce into its proxy inference var, instead.
+        let mut to_ty = to_ty;
+        let mut _to = None;
```
[the remark this answers is not in the document]
Comment (ShoyuVanilla) [reply crates/hir-ty/src/infer/coerce.rs:281]: Wow, this is awesome. Thanks!

**Comment by Veykril** [comment]
Awesome! Thanks!
@bors r+

**Pull request merged**